### DISCUSSION TITLE
Cloud Speech-to-Text API C++, Regarding initial silence timeout and end silence timeout

### DISCUSSION THREAD

**Discussion by gauravG2021** [googleapis/google-cloud-cpp#10302] [Q&A] [answered]
Hi,
How can we set 

1. Initial Silence Timeout: The amount of silence at the beginning of a recognition before the Speech Service stops recognizing speech.
3. End Silence Timeout: The amount of silence after speech before the Speech Service stops recognizing speech.

Thanks
Gaurav

**Comment by coryan** [comment, marked as answer]
I think these features are only available with Speech V2.  They are both documented here:

https://cloud.google.com/speech-to-text/v2/docs/reference/rpc/google.cloud.speech.v2#voiceactivitytimeout

Assuming you are using [`AsyncStreamingRecognize()`](https://googleapis.dev/cpp/google-cloud-speech/HEAD/classgoogle_1_1cloud_1_1speech__v2_1_1SpeechClient.html#af895455acad2eca141e5854f17817d6d) you need to fill the right fields in the `StreamingRecognizeRequest` message:

https://cloud.google.com/speech-to-text/v2/docs/reference/rpc/google.cloud.speech.v2#streamingrecognizerequest

That is, something like:

```cc
google::cloud::speech::v2::StreamingRecognize request;
auto& features = *request.mutable_streaming_config()->mutable_streaming_features();
features.set_enable_voice_activity_events(true);
auto& va = features.mutable_voice_activity_timeout();
va.mutable_speech_start_timeout().set_seconds(5);
va.mutable_speech_end_timeout().set_seconds(10);
```

Note that only the first `request` message may contain a streaming_config:

https://github.com/googleapis/googleapis/blob/67b2d7c2fb11188776bc398f02c67fccd8187502/google/cloud/speech/v2/cloud_speech.proto#L1106-L1111

You can specify sub-second timeouts too (see [Duration](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Duration)).

In general, the service documentation is the best way to find this type of information. Either in the [service guides](https://cloud.google.com/speech-to-text/v2/docs/voice-activity-events) or the [`.proto`](https://github.com/googleapis/googleapis/blob/master/google/cloud/speech/v2/cloud_speech.proto) file.

**Reply by gauravG2021** [reply]
Hi coryan,

if we set mutable_speech_end_timeout as 7 second.
"va.mutable_speech_end_timeout().set_seconds(7);"
Then how long we should wait for fetching the results.? i have to set a mximum duration  "X" for which i can wait for results, after the expire of x need to ask user to try again?

Also Speech V2 is in Pre-GA, when do we expect it be GA?

Thanks
Gaurav Gangwar

**Reply by coryan** [reply]
> Then how long we should wait for fetching the results.? i have to set a mximum duration "X" for which i can wait for results, after the expire of x need to ask user to try again?

Sorry, but I am not able to help.  I have no idea what the expected response time is for this service.  At a guess, I would wait for a few seconds, measure the performance in real-time, and set the timeout to something like the estimated 99.9% of the latency.


> Also Speech V2 is in Pre-GA, when do we expect it be GA?

I cannot help there.  I am not aware of the product plans for Speech V2.  If you have an account manager, or support, maybe you can get in touch with the product manager(s) for Speech V2.  Otherwise, maybe try to use one of the support channels listed here:

https://cloud.google.com/speech-to-text/docs/getting-support
